sdk/identity/azure-identity/src/azure_cli_credential.cpp

```diff
@@ -0,0 +1,654 @@
+      std::string::size_type& bytesRead,
+      bool& willHaveMoreData);
+};
+
+class ShellProcess final {
+private:
+#if defined(AZ_PLATFORM_WINDOWS)
+  HANDLE m_processHandle = NULL;
+#else
+  std::vector<char*> m_argv;
+  std::vector<char> m_argvValues;
+
+  std::vector<char*> m_envp;
+  std::vector<char> m_envpValues;
+
+  posix_spawn_file_actions_t m_actions = {};
+  int m_pid = -1;
+#endif
+
+  ShellProcess(ShellProcess const&) = delete;
+  ShellProcess& operator=(ShellProcess const&) = delete;
+
+public:
+  ShellProcess(std::string const& command, OutputPipe& outputPipe);
+  ~ShellProcess();
+
+  void Terminate();
+};
+
+std::string RunShellCommand(
+    std::string const& command,
+    DateTime::duration timeout,
+    Context const& context)
+{
+  // Use steady_clock so we're not affected by system time rewinding.
+  auto const terminateAfter = std::chrono::steady_clock::now()
+      + std::chrono::duration_cast<std::chrono::steady_clock::duration>(timeout);
+
+  std::string output;
+
+  OutputPipe pipe;
+  ShellProcess shellProcess(command, pipe);
+
+  // Typically token json is just a bit less than 2KiB.
+  // The best buffer size is the one that lets us to read it in one go.
+  // (Should it be smaller, we will succeed as well, it'll just take more iterations).
+  std::vector<std::string::value_type> processOutputBuf(2 * 1024);
+
+  auto willHaveMoreData = true;
+  do
+  {
+    // Check if we should terminate
+    {
+      if (context.IsCancelled())
+      {
+        shellProcess.Terminate();
+        throw std::runtime_error("Context was cancelled before Azure CLI process was done.");
+      }
+
+      if (std::chrono::steady_clock::now() > terminateAfter)
+      {
+        shellProcess.Terminate();
+        throw std::runtime_error("Azure CLI process took too long to complete.");
+      }
+    }
+
+    std::string::size_type bytesRead = 0;
+    if (pipe.NonBlockingRead(processOutputBuf, bytesRead, willHaveMoreData))
+    {
+      output.insert(output.size(), processOutputBuf.data(), bytesRead);
+    }
+    else if (willHaveMoreData)
+    {
+      std::this_thread::sleep_for(std::chrono::milliseconds(200)); // Value has no special meaning.
+    }
+  } while (willHaveMoreData);
+
+  return output;
+}
+
+#if defined(AZ_PLATFORM_WINDOWS)
+void ThrowIfApiCallFails(BOOL apiResult, std::string const& errMsg)
+{
+  // LCOV_EXCL_START
+  if (!apiResult)
+  {
+    throw std::runtime_error(
+        errMsg + ": " + std::to_string(GetLastError())
+
+    );
+  }
+  // LCOV_EXCL_STOP
+}
+#else
+void ThrowIfApiCallFails(int apiResult, std::string const& errMsg)
+{
+  // LCOV_EXCL_START
+  if (apiResult != 0)
+  {
+    throw std::runtime_error(
+        errMsg + ": " + std::to_string(apiResult) + " (errno: " + std::to_string(errno) + ")");
+  }
+  // LCOV_EXCL_STOP
+}
+#endif
+
+OutputPipe::OutputPipe()
+{
+#if defined(AZ_PLATFORM_WINDOWS)
+  SECURITY_ATTRIBUTES pipeSecurity = {};
+  ZeroMemory(&pipeSecurity, sizeof(decltype(pipeSecurity)));
```

Comment:
The ZeroMemory here is unnecessary, the initializer at line 282 does that already.